### DISCUSSION TITLE
Vercel Postgres 'missing_connection_string' localhost

### DISCUSSION THREAD

**Discussion by denisgolubovic** [vercel/community#2350] [Help] [answered]
### Summary

Trying out the new Postgres feature and I have created a simple test project with a Table in my Storage dashboard.

Pulled the environment variables and running it locally with next dev.

Running this insert

````typescript
import {sql} from "@vercel/postgres";
import {Post} from "@/interfaces/Post";

`export async function createPost(post: Post): Promise<void> {
    try {
        await sql`
            INSERT INTO Posts (id, title, content, author, date)
            VALUES (${post.id}, ${post.title}, ${post.content}, ${post.author}, ${post.date});
        `;
    } catch (e: any) {
        console.log('Could not insert post into the database');
        throw e;
    }
}`
````

Results in the following error:

````text
Failed to create post: VercelPostgresError: VercelPostgresError - 'missing_connection_string': You did not supply a 'connectionString' and no 'POSTGRES_URL' env var was found.
````

the .env.local is setup correct and i can log these with correct results:

````typescript
console.log({
      POSTGRES_URL: process.env.POSTGRES_URL,
      POSTGRES_URL_NON_POOLING: process.env.POSTGRES_URL_NON_POOLING
});
````

------EDIT---------
This on the other hand works.
````typescript
export async function fetchPosts() {
    try {
        const {rows, fields} = await sql`SELECT * FROM Posts;`;
        return rows as Post[];
    } catch (e: any){
       console.log("Could not fetch posts from database");
        return [];
    }
}
````

**Comment by amyegan** [comment, marked as answer]
Hi @denisgolubovic! This sounds similar to https://github.com/vercel/community/discussions/2304. Do any of the suggestions in that discussion work for you? 

Another thing to try, if you have not already, is using `vercel dev` instead of other local dev commands like `npm dev`. If any of that gets it working for you, please let me know what worked and which framework you're using so I can share it with the Storage engineering team.

If you still can't get it working, please share a minimal reproduction so we can look at the problem in more detail 🙏

**Reply by kevinclelland** [reply]
I had the same issue and using vercel dev instead of npm dev worked. Thank you

**Reply by JamesForan** [reply]
This worked for me also, using SvelteKit

**Reply by sojinsamuel** [reply]
Worked for me.  was facing issue on my nextjs learn chapters. Thnx for the fix

**Reply by raqueldebiase** [reply]
Thanks !!! This worked for me too!

**Comment by lkenny97** [comment]
In case someone makes a small mistake that leads them into an hour long journey to end up here:

My server action was connecting to Vercel Postgres. It was in a separate file. I forgot to put `"use server"` on top of the file. Obviously I wouldn't have postgres key on client side :D

**Reply by jacobhq** [reply]
Thank you, you saved me hours of time!

**Reply by MichalBastrzyk** [reply]
> In case someone makes a small mistake that leads them into an hour long journey to end up here:
> 
> My server action was connecting to Vercel Postgres. It was in a separate file. I forgot to put `"use server"` on top of the file. Obviously I wouldn't have postgres key on client side :D

In my case it was the two spaces instead of one in "use server" directive -_-

**Reply by MaxHiit** [reply]
> > In case someone makes a small mistake that leads them into an hour long journey to end up here:
> > My server action was connecting to Vercel Postgres. It was in a separate file. I forgot to put `"use server"` on top of the file. Obviously I wouldn't have postgres key on client side :D
> 
> In my case it was the two spaces instead of one in "use server" directive -_-

The same for me

**Comment by knowcart** [comment]
Apparently just making it .env instead of .env.local works.

**Reply by swlittles** [reply]
This works! Thanks for saving me an hour of troubleshooting! Wish Vercel could review their tutorials and fix issues with their instructions.

**Reply by walter-kai** [reply]
if you renamed it from .env.example, then you're reading from the same guide 😀

**Comment by jacobhq** [comment]
> In case someone makes a small mistake that leads them into an hour long journey to end up here:
> 
> My server action was connecting to Vercel Postgres. It was in a separate file. I forgot to put `"use server"` on top of the file. Obviously I wouldn't have postgres key on client side :D

This error is thrown when using your database client side, and is difficult to find because the error is not that descriptive. It is often abstracted into other files, making it even harder to find! Some clues this may be your issue:
- You don't see the error in the server console
- The error occurred after first render
- Calls to the database work elsewhere

@amyegan Please could you share this with the storage team, I think we may need a better error message! I'm happy to provide a repro is necessary.

Thanks everyone in this thread - you saved me lots of time! :)

**Reply by amyegan** [reply]
Thanks for flagging this! I opened a new ticket for the team to look into what we could do to help. 🙏

**Reply by jacobhq** [reply]
Thank you! Happy to provide any more info if you need it :)

**Reply by JackYinpei** [reply]
I think the method given here is the real cause of this error.
Thank you! Buddha blesses you！
Good people have good karma！

**Comment by AJR07** [comment]
For all **vite** users out there, you might want to check out the [Vercel Documentation on this issue](https://github.com/vercel/storage/tree/main/packages/postgres#a-note-for-vite-users), pertaining to how vite does not expose environment variables not ending in `VITE_`, so vercel would not be able to read it.
